### ISSUE TITLE
Arrow Core 1.2.0-RC breaks any Kapt task when using Raise/Effect

### ISSUE THREAD

**Issue by miikeat** [arrow-kt/arrow#3041]
We were already using 1.1.6-alpha.65 in one of our projects and tested the new features but when we wanted to switch to 1.2.0-RC our poject stopped building with the following error:

```
ExampleImpl.java:12: error: cannot access Raise
    public final kotlin.jvm.functions.Function2<arrow.core.raise.Raise<? super MyError>, kotlin.coroutines.Continuation<? super java.lang.String>, java.lang.Object> example() {
                                                                ^
  bad class file: /Users/kaiser/.gradle/caches/modules-2/files-2.1/io.arrow-kt/arrow-core-jvm/1.2.0-RC/815591a96f11786ae0ff746a4931fee9c0e46ac4/arrow-core-jvm-1.2.0-RC.jar(/arrow/core/raise/Raise.class)
    undeclared type variable: T
    Please remove or make sure it appears in the correct subdirectory of the classpath.
``` 

I created a minimal reproducable Example here: https://github.com/miikeat/ArrowKaptIssue

I narrowed the Issue down to the changes in alpha.73 -> alpha.74 (Diff: https://github.com/arrow-kt/arrow/compare/1.1.6-alpha.73...1.1.6-alpha.74)

And as you can see in build.gradle.kts in that project it only happens when a kapt(...) dependency is present.

```
dependencies {
    // as soon as any of the following two lines are not commented out, the above error appears with alpha.74 or higher
    kapt("com.datastax.oss:java-driver-mapper-processor:4.15.0")
    //kapt("androidx.room:room-compiler:2.5.0")

    // Diff between alpha 73 & 74: https://github.com/arrow-kt/arrow/compare/1.1.6-alpha.73...1.1.6-alpha.74
    //implementation("io.arrow-kt:arrow-core:1.1.6-alpha.73") // Last working Version
    //implementation("io.arrow-kt:arrow-core:1.1.6-alpha.74") // First Version with that Issue
    implementation("io.arrow-kt:arrow-core:1.2.0-RC")
    testImplementation(kotlin("test"))
}
```

Maybe thats a bug of Kapt but as it worked until version 1.1.6-alpha.73 I thought its probably better to post this first here!

**Comment by nomisRev** [comment]
Hey @miikeat,

Thank you so much for the reproducible, and pin pointing it to the exact alpha’s 🙌
This is great, will look into it asap.

**Comment by bddckr** [comment]
I've been unable to use 1.2.0-RC for this exact reason. Is there anything we can help with to identify the root cause further and fix this?

**Comment by serras** [comment]
@franciscodr and I have been doing some thorough investigation this morning. It turns out that the compiler is generating _wrong_ bytecode for this specific scenario. The default implementation of `bindAll` for `NonEmptyList` (figuring this out would have been impossible without the prior investigation from @miikeat) generates a wrong type, as seen here in the decompiler.

<img width="1116" alt="Captura de pantalla 2023-06-02 a las 10 36 21" src="https://github.com/arrow-kt/arrow/assets/309334/2447dca8-796d-4c79-9095-06bf051effb3">

Note that the type argument is `<A>`, but afterward `<T>` is used! This coincides with the error message about an unbound variable `T` in the `.class` file. This means the error is in the compiler, but we're now looking at forms to force the compiler to generate the right code.

**Comment by serras** [comment]
Fixed by #3062
